text/0000-subslice-offset.md

```diff
@@ -0,0 +1,344 @@
+(...)
+v.truncate(end_file_stem.wrapping_sub(start));
+```
+
+With the proposed functions, it'd look like this:
+
+```rust
+let file_stem_range = os_str_as_u8_slice(&self.inner)
+    .range_of(file_stem)
+    .expect("file_stem() gave a slice outside of the path");
+(...)
+v.truncate(file_stem_range.end);
+```
+
+[offset_example]: https://github.com/rust-lang/rust/blob/18ae175d60039dfe2d2454e8e8099d3b08039862/src/libstd/path.rs#L1372-L1375
+
+#### Use Case Example 2: Increase Efficiency
+
+For this example, take this struct representing some token in a tokenized source
+file:
+
+```rust
+struct Token<'a> {
+  (...)
+  source: &'a str, // the exact source of this token
+  offset_in_source_file: usize, // for error reporting
+}
+```
+
+If the `source` always points into the full `String` contents of the file,
+keeping `offset_in_source_file` is a waste of space, as that information is
+already encoded in the memory address stored in `source`:
+`file_contents.range_of(token.source)`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## index_of
+
+```rust
+impl<T> [T] {
+    pub fn index_of(&self, element: &T) -> Option<usize>
+}
+```
+
+Returns the index of the element that a reference refers to.
+
+If the given reference points inside the slice, returns the index of
+the element it refers to. If the reference does not refer to within the
+slice, `None` is returned instead.
+
+If an index is returned, it is less than `self.len()`.
+
+Note that this does not look at the value, but only at the reference.
+If you want to find the index of an element equal to a given value, use
+`iter().position()` instead.
+
+```rust
+let a = [0; 5];
+
+assert_eq!(a.index_of(&a[2]), Some(2));
+assert_eq!(a.index_of(&3), None);
+```
+
+## range_of
+
+```rust
+impl<T> [T] {
+    pub fn range_of(&self, subslice: &[T]) -> Option<Range<usize>>
+}
+```
+
+Returns the range referred to by a subslice.
+
+If the given slice falls entirely within this slice, returns the range
+of indices the subslice refers to. If the given slice is not a subslice
+of this slice, `None` is returned.
+
+If a range is returned, both ends are less than or equal to `self.len()`.
+
+Note that this does not look at the contents of the slice, but only at
+the memory addresses.
+
+```rust
+let a = [0; 5];
+
+assert_eq!(a.range_of(&a[2..5]), Some(2..5));
+assert_eq!(a.range_of(&[7, 8, 9]), None);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The functions below are added to `impl<T> [T] { .. }` in `src/libcore/slice/mod.rs`,
+with doc comments based on the [Guide-level explanation][guide-level-explanation] above.
+
+```rust
+    pub fn index_of(&self, element: &T) -> Option<usize> {
+        let element = element as *const _;
+        let range = self.as_ptr_range();
+        if range.contains(&element) {
+            unsafe { Some(element.offset_from(range.start) as usize) }
```

Comment:
This implementation panics for ZSTs (but would otherwise be unsound anyways). The requirements of `offset_from` state:

>  Both the starting and other pointer must be either in bounds or one byte past the end of the same allocated object. Note that in Rust, every (stack-allocated) variable is considered a separate allocated object.
> [...]
> Panics
> This function panics if T is a Zero-Sized Type ("ZST").

References to ZSTs are always dangling pointers and thus `offset_from` never valid as they never point into any allocation. The non-ZST case could be improved with an explicit comment detailing soundness justifications.

Also note that even if having a special case implementation, slices of ZST will return `Some(0)` in many more cases than the caller might realize, sometimes depending on optimization levels.

```
let not_an_allocation = vec![(); 16];
let some_new = &();
// Entirely indetermined, might hold or not depending on compiler
assert!(not_an_allocation.index_of(&some_new).is_some());
```